File: javascript/src/trees/bTree.js
/** Trees — B-Tree */

/**
 * B-tree with minimum degree t (each node has at most 2t-1 keys).
 * search/insert O(log n).
 */
class BNode {
  constructor() {
    this.leaf = false;
    this.keys = [];
    this.children = [];
  }
}

class BTree {
  /**
   * Creates a B-tree with minimum degree t (t >= 2).
   * Time: O(1); Space: O(1).
   */
  constructor(t) {
    if (t < 2) {
      throw new Error('t must be >= 2');
    }
    this._minDegree = t;
    this._root = new BNode();
    this._root.leaf = true;
  }

  /** Searches for a key. Time: O(log n); Space: O(1). */
  search(key) {
    return this._search(this._root, key);
  }

  /** Inserts a key. Time: O(log n); Space: O(1). */
  insert(key) {
    const r = this._root;
    if (r.keys.length === 2 * this._minDegree - 1) {
      const newRoot = new BNode();
      newRoot.leaf = false;
      newRoot.children.push(this._root);
      this._splitChild(newRoot, 0);
      this._root = newRoot;
    }
    this._insertNonFull(this._root, key);
  }

  _search(node, key) {
    let i = 0;
    while (i < node.keys.length && key > node.keys[i]) {
      i += 1;
    }
    if (i < node.keys.length && key === node.keys[i]) {
      return true;
    }
    if (node.leaf) {
      return false;
    }
    return this._search(node.children[i], key);
  }

  _insertNonFull(node, key) {
    let i = node.keys.length - 1;
    if (node.leaf) {
      node.keys.push(key);
      while (i >= 0 && key < node.keys[i]) {
        node.keys[i + 1] = node.keys[i];
        i -= 1;
      }
      node.keys[i + 1] = key;
    } else {
      while (i >= 0 && key < node.keys[i]) {
        i -= 1;
      }
      i += 1;
      if (node.children[i].keys.length === 2 * this._minDegree - 1) {
        this._splitChild(node, i);
        if (key > node.keys[i]) {
          i += 1;
        }
      }
      this._insertNonFull(node.children[i], key);
    }
  }

  _splitChild(parent, index) {
    const full = parent.children[index];
    const newNode = new BNode();
    newNode.leaf = full.leaf;
    const t = this._minDegree;
    for (let j = 0; j < t - 1; j += 1) {
      newNode.keys.push(full.keys[j + t]);
    }
    if (!full.leaf) {
      for (let j = 0; j < t; j += 1) {
        newNode.children.push(full.children[j + t]);
      }
      full.children.splice(t);
    }
    const mid = full.keys[t - 1];
    full.keys.splice(t - 1);
    parent.keys.splice(index, 0, mid);
    parent.children.splice(index + 1, 0, newNode);
  }
}

module.exports = { BTree };

File: java/src/main/java/com/dsa/examples/trees/BTree.java
// Trees — B-Tree

package com.dsa.examples.trees;

import java.util.ArrayList;
import java.util.List;

/**
 * B-tree with minimum degree t (each node has at most 2t-1 keys).
 * search/insert O(log n).
 */
public class BTree {

    private final int minDegree;
    private Node root;

    private static class Node {
        boolean leaf;
        List<Integer> keys = new ArrayList<>();
        List<Node> children = new ArrayList<>();
    }

    /**
     * Creates a B-tree with minimum degree {@code t} (t >= 2).
     * Time: O(1); Space: O(1).
     */
    public BTree(int t) {
        if (t < 2) {
            throw new IllegalArgumentException("t must be >= 2");
        }
        this.minDegree = t;
        root = new Node();
        root.leaf = true;
    }

    /**
     * Searches for a key.
     * Time: O(log n); Space: O(1).
     */
    public boolean search(int key) {
        return search(root, key);
    }

    /**
     * Inserts a key.
     * Time: O(log n); Space: O(1).
     */
    public void insert(int key) {
        Node r = root;
        if (r.keys.size() == 2 * minDegree - 1) {
            Node newRoot = new Node();
            newRoot.leaf = false;
            newRoot.children.add(root);
            splitChild(newRoot, 0);
            root = newRoot;
        }
        insertNonFull(root, key);
    }

    private boolean search(Node node, int key) {
        int i = 0;
        while (i < node.keys.size() && key > node.keys.get(i)) {
            i++;
        }
        if (i < node.keys.size() && key == node.keys.get(i)) {
            return true;
        }
        if (node.leaf) {
            return false;
        }
        return search(node.children.get(i), key);
    }

    private void insertNonFull(Node node, int key) {
        int i = node.keys.size() - 1;
        if (node.leaf) {
            node.keys.add(key);
            while (i >= 0 && key < node.keys.get(i)) {
                node.keys.set(i + 1, node.keys.get(i));
                i--;
            }
            node.keys.set(i + 1, key);
        } else {
            while (i >= 0 && key < node.keys.get(i)) {
                i--;
            }
            i++;
            if (node.children.get(i).keys.size() == 2 * minDegree - 1) {
                splitChild(node, i);
                if (key > node.keys.get(i)) {
                    i++;
                }
            }
            insertNonFull(node.children.get(i), key);
        }
    }

    private void splitChild(Node parent, int index) {
        Node full = parent.children.get(index);
        Node newNode = new Node();
        newNode.leaf = full.leaf;
        int t = minDegree;
        for (int j = 0; j < t - 1; j++) {
            newNode.keys.add(full.keys.get(j + t));
        }
        if (!full.leaf) {
            for (int j = 0; j < t; j++) {
                newNode.children.add(full.children.get(j + t));
            }
            full.children.subList(t, full.children.size()).clear();
        }
        int mid = full.keys.get(t - 1);
        full.keys.subList(t - 1, full.keys.size()).clear();
        parent.keys.add(index, mid);
        parent.children.add(index + 1, newNode);
    }
}
